### PR TITLE
Fix extended tests

### LINKED ISSUE #11903: Multiple test fails
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin_extended/793/

	Test Result (11 failures / +10)

    Extended.[k8s.io] Garbage collector [Feature:GarbageCollector] should delete pods created by rc when not orphaning
    Extended.[networking] OVS generic should add and remove flows when pods are added and removed
    Extended.[imageapis] openshift limit range admission should deny a push of built image exceeding openshift.io/Image limit
    Extended.[k8s.io] Deployment RollingUpdateDeployment should scale up and down in the right order
    Extended.[networking] OVS generic should add and remove flows when nodes are added and removed
    Extended.[k8s.io] Federation secrets [Feature:Federation12] Secret objects should be created and deleted successfully
    Extended.[image_ecosystem][jenkins] schedule jobs on pod slaves use of jenkins with kubernetes plugin by creating slave from existing builder and adding it to Jenkins master
    Extended.[k8s.io] Deployment deployment should label adopted RSs and pods
    Extended.[k8s.io] Deployment RollingUpdateDeployment should delete old pods and create new ones
    Extended.[k8s.io] Deployment RecreateDeployment should delete old pods and create new ones

### PR DESCRIPTION
**Pull request by ncdc**
- allow deployment controller to update pods
- disable kube garbage collector tests since the kube garbage collector isn't enabled in origin yet
- excluded 1 federation test that wasn't caught by our existing filter
- skip networking OVS tests in core.sh because they need a special network configuration that's set up by test/extended/networking.sh

Fixes #11903

### CODE REVIEW

**Comment by ncdc** [comment]
@deads2k @liggitt @pweil- @mfojtik @kargakis @soltysh @smarterclayton @bparees FYI, working through extended test failures. I'll be adding more commits as I find more things.

**Comment by bparees** [comment]
@oatmealraisin may be of interest to you.

**Comment by mfojtik** [comment]
@ncdc ❤️

**Comment by deads2k** [comment]
How long was this broken?

Also, you're going to fail a unit test which give instructions for updating the yaml that corresponds to the role.

**Comment by ncdc** [comment]
@deads2k yeah this is a WIP. I'm not going to run it through jenkins until I feel like most or all of the issues are resolved. Extended tests have been broken for a while.

File: pkg/cmd/server/bootstrappolicy/infra_sa_policy.go [hunk 1]
```diff
@@ -230,7 +230,7 @@ func init() {
 				},
 				{
 					APIGroups: []string{""},
-					Verbs:     sets.NewString("get", "list", "watch"),
+					Verbs:     sets.NewString("get", "list", "watch", "update"),
```
Comment (liggitt) [inline pkg/cmd/server/bootstrappolicy/infra_sa_policy.go]: this really seems like controlling the wrong layer :(
do we want to open an issue about it upstream?

Comment (ncdc) [inline pkg/cmd/server/bootstrappolicy/infra_sa_policy.go]: I thought you'd say that. How would you prefer to do it?

Comment (ncdc) [inline pkg/cmd/server/bootstrappolicy/infra_sa_policy.go]: cc @kargakis

Comment (kargakis) [inline pkg/cmd/server/bootstrappolicy/infra_sa_policy.go]: We need the permission to enable adoption of replica sets (pods may be relabeled).

Comment (kargakis) [inline pkg/cmd/server/bootstrappolicy/infra_sa_policy.go]: Do you think that a label or relabel subresource on the replica set that would make the replica set controller relabel itself and then the pods would make any sense? It's not possible now so we need to handle it in deployments.

**Comment by kargakis** [comment]
LGTM

File: pkg/cmd/server/bootstrappolicy/infra_sa_policy.go (hunk 1 above)
Comment (ncdc) [inline pkg/cmd/server/bootstrappolicy/infra_sa_policy.go]: I think a subresource is the way to go

**Comment by ncdc** [comment]
@kargakis do you think you can take over this work?

**Comment by kargakis** [comment]
> @kargakis do you think you can take over this work?

Not sure if I can make it to the rest of the test flakes today. Is there anything else that needs to be done for the deployment flakes?

**Comment by ncdc** [comment]
All the deployment test failures pass with the added "update pods" permission for the deployment controller. I believe the other 3 commits should address most if not all of the remaining actual failures (not flakes).

**Comment by ncdc** [comment]
[testextended][extended:core]

**Comment by ncdc** [comment]
[testextended][extended:core]

**Comment by stevekuznetsov** [comment]
While we're here we might want to lint these lists... a lot of the literal whitelist for conformance matches nothing...
```sh
~$ cat /tmp/test.sh 
#!/bin/bash

source /home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh

for test in "${CONFORMANCE_TESTS[@]}"; do
	matches="$(grep -nrE "${test}" /home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended /home/skuznets/Documents/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test)"
	if ! grep -q "It(" <<<"${matches}" && ! grep -q "g.Describe(" <<<"${matches}" && ! grep -q "framework.KubeDescribe" <<<"${matches}" ; then
		echo "Test \`${test}\` matches no actual tests!"
		echo "Matches found: "
		echo "${matches:-none}"
		echo
	fi
done
```

```
~$ /tmp/test.sh 
Test `Services.*NodePort` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:287:	"Services.*NodePort"
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/networking.sh:19:NETWORKING_E2E_FOCUS="${NETWORKING_E2E_FOCUS:-etworking|Services should be able to create a functioning NodePort service|EmptyDir volumes should support \(root,0644,tmpfs\)}"
/home/skuznets/Documents/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/resource_quota.go:509:	hard[api.ResourceServicesNodePorts] = resource.MustParse("1")

Test `ResourceQuota should` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:288:	"ResourceQuota should"

Test `\[networking\] basic openshift networking` matches no actual tests!
Matches found: 
none

Test `DNS for pods for Hostname and Subdomain annotation` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:295:	"DNS for pods for Hostname and Subdomain annotation"

Test `PrivilegedPod should test privileged pod` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:296:	"PrivilegedPod should test privileged pod"

Test `Pods should support remote command execution` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:297:	"Pods should support remote command execution"

Test `Pods should support retrieving logs from the container` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:298:	"Pods should support retrieving logs from the container"

Test `Kubectl client Simple pod should support` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:299:	"Kubectl client Simple pod should support"

Test `Job should run a job to completion when tasks succeed` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:300:	"Job should run a job to completion when tasks succeed"

Test `\[images\]\[mongodb\] openshift mongodb replication` matches no actual tests!
Matches found: 
none

Test `\[job\] openshift can execute jobs controller` matches no actual tests!
Matches found: 
none

Test `\[volumes\] Test local storage quota FSGroup` matches no actual tests!
Matches found: 
none

Test `test deployment should run a deployment to completion` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:304:	"test deployment should run a deployment to completion"

Test `Clean up pods on node kubelet` matches no actual tests!
Matches found: 
/home/skuznets/Documents/go/src/github.com/openshift/origin/test/extended/setup.sh:307:	"Clean up pods on node kubelet"
```

**Comment by ncdc** [comment]
Still need to fix:

```
• Failure in Spec Setup (BeforeEach) [5.321 seconds]
[image_ecosystem][jenkins] schedule jobs on pod slaves
/data/src/github.com/openshift/origin/test/extended/image_ecosystem/kubernetes_plugin.go:169
  use of jenkins with kubernetes plugin
  /data/src/github.com/openshift/origin/test/extended/image_ecosystem/kubernetes_plugin.go:168
    by creating slave from existing builder and adding it to Jenkins master [BeforeEach]
    /data/src/github.com/openshift/origin/test/extended/image_ecosystem/kubernetes_plugin.go:167

    Expected error:
        <*os.PathError | 0xc82171b050>: {
            Op: "open",
            Path: "examples/jenkins/master-slave/jenkins-master-template.json",
            Err: 0x2,
        }
        open examples/jenkins/master-slave/jenkins-master-template.json: no such file or directory
    not to have occurred

    /data/src/github.com/openshift/origin/test/extended/image_ecosystem/kubernetes_plugin.go:22
```

This passes for me locally, so something must be weird with the devenv:

```
• Failure [15.623 seconds]
[imageapis] openshift limit range admission
/data/src/github.com/openshift/origin/test/extended/imageapis/limitrange_admission.go:224
  should deny a push of built image exceeding limit on openshift.io/images resource [It]
  /data/src/github.com/openshift/origin/test/extended/imageapis/limitrange_admission.go:122

  Expected error:
      <*errors.errorString | 0xc8209c1880>: {
          s: "Failed to match expected \"(?i)requested access to the resource is denied|failed to push image: denied|^denied$\" in: \"2016-11-16T16:54:36.783959000Z Receiving source from STDIN as archive ...\\n2016-11-16T16:54:37.573859000Z Pulling image scratch ...\\n2016-11-16T16:54:37.575120000Z error: build error: failed to pull image: API error (500): {\\\"message\\\":\\\"'scratch' is a reserved name\\\"}\"",
      }
      Failed to match expected "(?i)requested access to the resource is denied|failed to push image: denied|^denied$" in: "2016-11-16T16:54:36.783959000Z Receiving source from STDIN as archive ...\n2016-11-16T16:54:37.573859000Z Pulling image scratch ...\n2016-11-16T16:54:37.575120000Z error: build error: failed to pull image: API error (500): {\"message\":\"'scratch' is a reserved name\"}"
  not to have occurred

  /data/src/github.com/openshift/origin/test/extended/imageapis/limitrange_admission.go:87
```

**Comment by bparees** [comment]
@gabemontero can look at the jenkins failure, i think it's already on our radar because we run our extended tests nightly.

**Comment by ncdc** [comment]
@bparees thanks, that would be helpful. It's not a flake.

**Comment by gabemontero** [comment]
@bparees @ncdc the kubernetes_plugin.go failure noted above does not make sense to me, given that file still exists in origin's master branch:

https://github.com/openshift/origin/blob/master/examples/jenkins/master-slave/jenkins-master-template.json

And I've seen that test pass as recently as https://ci.openshift.redhat.com/jenkins/job/origin_extended_image_tests/754/consoleFull and that file has not changed since Oct 31.

side bar: that test is going away once https://github.com/openshift/origin/pull/11844 merges, though that should have no bearing on why the above failures are occuring, and why a file which does exist is found in our nightly runs but not in these runs.

**Comment by gabemontero** [comment]
@bparees I have been scanning a few of our nightly runs and have seen no indication of this file not found issue.

**Comment by gabemontero** [comment]
@ncdc @bparees could these extended test runs here possibly be occurring under a different directory (and hence the relative path in the test case is invalid) ?

**Comment by ncdc** [comment]
That would be my guess

**Comment by stevekuznetsov** [comment]
>  I have been scanning a few of our nightly runs

@gabemontero do we not have historical jUnit for it?

**Comment by ncdc** [comment]
Only 3 test failures now!

**Comment by ncdc** [comment]
@oatmealraisin and @gabemontero are looking into the 3 remaining failures (2 tests failing to pull scratch, 1 file not found).

Any objections to merging this?

[test]

**Comment by kargakis** [comment]
> Any objections to merging this?

@mfojtik we definitely want the first commit.

**Comment by ncdc** [comment]
Corresponding PR for 1.4: #11942

**Comment by gabemontero** [comment]
Comparing https://ci.openshift.redhat.com/jenkins/job/origin_extended_image_tests/configure with https://ci.openshift.redhat.com/jenkins/job/test_pr_origin_extended/configure there are differences in how the extended tests are launched.

Assuming those differences result in changes in the relative dir, the old kubernetes_plugin.go test builds the file path with [`filepath.Join`](https://github.com/openshift/origin/blob/master/test/extended/image_ecosystem/kubernetes_plugin.go#L57), where as our plugin.go tests leverage [`FixturePath`](https://github.com/openshift/origin/blob/master/test/extended/image_ecosystem/plugin.go#L335), which ultimately looks and the env var `EXTENDED_TEST_PATH`, which I'm guessing is properly set irregardless of where the tests start from.

@bparees - can we just wait until https://github.com/openshift/origin/pull/11844 merges and kubernetes_plugin.go goes away?  Or do you want kubernetes_plugin.go updated in the interim?

**Comment by bparees** [comment]
@gabemontero it can wait, but i'm surprised there's a meaningful difference between the two test jobs.  the only difference should be the focus being passed to the bucket, so we should consider reconciling the jobs.

**Comment by stevekuznetsov** [comment]
> so we should consider reconciling the jobs.

Quite a lot of our jobs are almost identical with small config changes, if we had a way to describe them that way we would be much better off.

**Comment by ncdc** [comment]
Sounds like it's time to move to jjb

**Comment by stevekuznetsov** [comment]
A way that doesn't involve having one job and passing it parameters, since that is not very nice for aggregate data, etc.

**Comment by ncdc** [comment]
Fixing the unit test failure from my policy change...

File: test/extended/setup.sh [hunk 2]
```diff
@@ -214,6 +214,9 @@ readonly EXCLUDED_TESTS=(
 
 	"\[Feature:Performance\]"
 
+	# not enabled in Origin yet
+	"\[Feature:GarbageCollector\]"
```
Comment (deads2k) [inline test/extended/setup.sh]: Let's get an issue for this in 1.5.  Some controllers are starting to use it.

**Comment by deads2k** [comment]
lgtm

**Comment by ncdc** [comment]
[merge]

**Pull request merged**